_posts/2022-09-XX-rate-limiting-web-applications.markdown

```diff
@@ -0,0 +1,465 @@
+* Partition per endpoint
+* Partition per IP address
+  * Keep in mind users may be sharing an IP address, e.g. when behind a NAT/CGNAT/proxy
+* Partition per user
+  * Keep in mind you may have anonymous users, how will you distinguish those?
+* Partition per session
+  * What if your user starts a new session for every request?
+  * What if your user makes use of multiple devices with separate sessions?
+* Partition per browser
+* Partition per header (e.g. `X-Api-Token`), ...
+
+Also here, "it depends" on your application. A global rate limit per IP address may work for your application.
+More complex applications may need a combination of these, e.g. per-endpoint rate limiting combined with the current user.
+
+### Decide on exceptions
+
+Should rate limiting apply to all requests? Well... yes! We already discussed all endpoints in your application should be rate limited.
+
+A better question would be whether the same limits should apply for all types of users. As usual, the answer to this question will depend on your application.
+There is no silver bullet, but here are some examples to think about.
+
+Good candidates to have different rate limits in place:
+* Your automated monitoring - the last thing you want is nightly PagerDuty alerts because of your monitoring system being rate limited.
+  * Counter point: maybe you do want to have a rate limit in place, so your monitoring can check rate limits are enabled?
+* Your admin/support team - your support team may make a lot of requests to your application to help out users, so it's best to not get in their way.
+  * Counter point: maybe a rate limit does make sense, so a disgruntled employee can't go and scrape lots of data or add swear words into lots of places with an automated script.
+* Web crawlers - your marketing folks won't be happy if your app is not visible in search engines!
+  * Counter point: there are aggressive crawlers, and you also don't want them to get in the way of your users. There are some `robots.txt` entries many spiders respect, but a rate limit could be needed.
+* "Datacenter IP ranges" - If you have a mobile app targeted at consumers, does traffic coming from AWS, Azure and other big hosters make sense? If you expect mostly "residential" and mobile traffic, perhaps you can reduce automated traffic from other sources with a more strict rate limit. There's a list of [datacenter IP ranges](https://github.com/client9/ipcat) that you can use for this.
+  * Counter point: VPN providers out there are often going to be in these IP ranges. Legitimate users may use datacenter IP addresses in those cases. Also at the time of writing, my dad's Starlink subscription runs over what looks like a Google Compute Engine IP address.
+
+Also here, measuring will help you make an informed decision. If you see excess traffic from web crawlers, a tighter rate limit may be needed. If you see your support folks unable to help users, maybe a less strict rate limit for them makes more sense.
```

Comment:
Use custom metrics to build dashboards on # of rate limiting actions kicking in to help during incident troubleshooting

Reply:
Took this one further in the text, good suggestion!